### PR TITLE
Initial 3D code

### PR DESCRIPTION
**Pull request by michalJur**
Added initial 3D code without changing much else

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -21,10 +21,17 @@
             "console": "integratedTerminal"
         },
         {
-            "name": "Simulator",
+            "name": "Simulator_2D",
```
Comment (piotrbartman) [inline .vscode/launch.json:24]: Remove this file from PR

File: deep_conmech/common/plotter/plotter_3d.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    ax.add_collection3d(
+        Poly3DCollection(
+            boundary_faces_nodes,
+            edgecolors=color,
+            linewidths=0.1,
+            facecolors=color,
+            alpha=0.2,
+        )
+    )
+    """
```
Comment (piotrbartman) [inline deep_conmech/common/plotter/plotter_3d.py:21]: remove unnecessary comment

File: deep_conmech/common/plotter/plotter_3d.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    ax,
+):
+    draw_base_arrows(ax, moved_base)
+
+    moved_boundary_faces_nodes = moved_nodes[boundary_faces]
+    moved_boundary_faces_centers = np.mean(moved_boundary_faces_nodes, axis=1)
+    plot_arrows(ax, starts=moved_boundary_faces_centers, vectors=boundary_normals)
+
+
+    '''
```
Comment (piotrbartman) [inline deep_conmech/common/plotter/plotter_3d.py:161]: remove unnecessary comment

File: deep_conmech/common/plotter/plotter_3d.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        ax=ax,
+    )
+
+    angles = np.array([[[0, -90], [0, 0]], [[30, -60], [90, 0]]])
+    distances = np.array([[10, 10], [11, 10]])
+    rows, columns, _ = angles.shape
+
+    fig = plt.figure()  # constrained_layout=True)
+    grid = fig.add_gridspec(nrows=rows, ncols=columns)
+    # , width_ratios=[1, 1.], height_ratios=[1., 1.])
```
Comment (piotrbartman) [inline deep_conmech/common/plotter/plotter_3d.py:107]: remove unnecessary comment

File: deep_conmech/common/plotter/plotter_3d.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    ax.w_zaxis.pane.fill = False
+
+    aspect = (12, 4, 4)
+    ax.set_box_aspect(aspect)
+
+    ax.set_xlim(-1, aspect[0] - 1)
+    ax.set_ylim(-1, aspect[1] - 1)
+    ax.set_zlim(-1, aspect[2] - 1)
+
+    # ax.set_xlabel("x", color="w")
```
Comment (piotrbartman) [inline deep_conmech/common/plotter/plotter_3d.py:56]: remove unnecessary comment

File: deep_conmech/common/plotter/plotter_3d.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+def plt_save(path, extension):
+    plt.savefig(
+        path,
+        transparent=False,
+        facecolor="#24292E",  # AAAAAA',
+        bbox_inches="tight",
+        pad_inches=0.0,
+        format=extension,
+        dpi=1000,  # 800 1200,
```
Comment (piotrbartman) [inline deep_conmech/common/plotter/plotter_3d.py:200]: remove unnecessary comment

File: deep_conmech/common/plotter/plotter_3d.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+        shifted_normalized_nodes = shifted_normalized_nodes + np.array([2.5, 0, 0])
+    
+
+
+def plt_save(path, extension):
+    plt.savefig(
+        path,
+        transparent=False,
+        facecolor="#24292E",  # AAAAAA',
```
Comment (piotrbartman) [inline deep_conmech/common/plotter/plotter_3d.py:196]: remove unnecessary comment

File: deep_conmech/graph/data/data_interpolation.py [hunk 8]
```diff
@@ -12,7 +12,7 @@ def weighted_mean_numba(v1, v2, scale):
 
 @njit
 def interpolate_point_numba(initial_point, corner_vectors, scale_x, scale_y):
-    min = [0., 0.]  #########
+    min = [0.0, 0.0]  #########
```
Comment (piotrbartman) [inline deep_conmech/graph/data/data_interpolation.py:15]: remove unnecessary comment

File: deep_conmech/graph/model.py [hunk 9]
```diff
@@ -97,7 +97,7 @@ def solve(self, setting, print_time=False):
             print("Graph solve time: ", time.time() - start)
 
         normalized_a = thh.to_np_double(normalized_a_cuda)
-        a = setting.rotate_from_upward(normalized_a)
+        a = setting.denormalize_rotate(normalized_a)
         return a
 
     ################
```
Comment (piotrbartman) [inline deep_conmech/graph/model.py:103]: remove unnecessary comment

File: deep_conmech/simulator/mesh/mesh_builders_2d.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
+from ctypes import ArgumentError
+# import dmsh
```
Comment (piotrbartman) [inline deep_conmech/simulator/mesh/mesh_builders_2d.py:2]: remove unnecessary comment

File: deep_conmech/simulator/mesh/mesh_builders_2d.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+    def edge_size(x):
+        return mesh_density  # + 0.1 * path.dist(x)
+
+    points_initial, cells_initial = dmsh.generate(geo, edge_size, tol=1.0e-10)
+    return optimesh.optimize_points_cells(
+        points_initial, cells_initial, "CVT (full)", 1.0e-10, 100
+    )
+
+###############################
```
Comment (piotrbartman) [inline deep_conmech/simulator/mesh/mesh_builders_2d.py:39]: remove unnecessary comment

File: deep_conmech/simulator/mesh/mesh_builders_helpers.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
+import numpy as np
+from deep_conmech.graph.data.data_interpolation import interpolate_point_numba
+from numba import njit
+
+
+@njit
+def random_corner_mesh_size(mesh_density):
+    scale = mesh_density * 0.8
+    random_data = np.random.rand(4)
+    # random_data = np.zeros(4) #random_data[1] = 1.
```
Comment (piotrbartman) [inline deep_conmech/simulator/mesh/mesh_builders_helpers.py:10]: remove unnecessary comment

File: deep_conmech/simulator/mesh/mesh_builders_helpers.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    corner_data = (random_data * 2.0 * scale) - scale
+    return 1.0 / (mesh_density + corner_data)
+
+    # z = np.sin(np.sqrt(x**2 + y**2))
```
Comment (piotrbartman) [inline deep_conmech/simulator/mesh/mesh_builders_helpers.py:14]: remove unnecessary comment

File: deep_conmech/simulator/setting/setting_mesh.py [hunk 14]
```diff
@@ -79,23 +79,40 @@ def move_boundary_points_to_start(
...
+
+def get_base_seed_indices(nodes):
+    dim = nodes.shape[1]
+    base_seed_indices = np.zeros((dim, 2), dtype=np.int64)
+    errors = np.zeros(dim)
+    for i in range(dim):
+        result = get_closest_to_axis(nodes, i)
+        errors[i] = result[0]
+        base_seed_indices[i] = result[1:].astype(np.int64)
+        # print(f"MIN ERROR for variable {i}: {errors[i]}")
```
Comment (piotrbartman) [inline deep_conmech/simulator/setting/setting_mesh.py:108]: remove unnecessary comment

File: deep_conmech/starter_simulator_3d.py [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+    boundary_internal_nodes = moved_nodes[boundary_internal_indices]
+
+    va = boundary_faces_nodes[...,1]-boundary_faces_nodes[...,0]
+    vb = boundary_faces_nodes[...,2]-boundary_faces_nodes[...,0]
+    vc = np.cross(va,vb)
+
+    boundary_normals = nph.normalize_euclidean_numba(va)
+    return boundary_normals
+
+#nph.elementwise_dot(vc, boundary_internal_nodes) > 0
```
Comment (piotrbartman) [inline deep_conmech/starter_simulator_3d.py:85]: remove unnecessary comment

File: deep_conmech/starter_simulator_3d.py [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+mean_initial_nodes = np.mean(initial_nodes, axis=0)
+normalized_initial_nodes = normalize_rotate(
+    initial_nodes - mean_initial_nodes, initial_base
+)
+
+
+edges_features_matrix, element_initial_volume = get_edges_features_matrix_numba(
+    elements, initial_nodes
+)
+# TODO: To tests - sum off slice for area and u == 1
```
Comment (piotrbartman) [inline deep_conmech/starter_simulator_3d.py:104]: Write test for that

**Review by piotrbartman — COMMENTED** [review]
Can you add at lest one simple regression test?

File: .vscode/launch.json (hunk 1 above)
Comment (michalJur) [reply .vscode/launch.json]: Done

File: deep_conmech/starter_simulator_3d.py (hunk 16 above)
Comment (michalJur) [reply deep_conmech/starter_simulator_3d.py]: Done

**Pull request merged**